src/lib/AudioUltra/Media/WaveformAudio.ts

```diff
@@ -1,38 +1,59 @@
+import { Events } from '../Common/Events';
+import { AudioDecoder } from './AudioDecoder';
+
 export interface WaveformAudioOptions {
-  volume: number;
-  muted: boolean;
-  rate: number;
+  src?: string;
+  volume?: number;
+  muted?: boolean;
+  rate?: number;
+}
+
+interface WaveformAudioEvents {
+  decodingProgress: (chunk: number, total: number) => void;
 }
 
-export class WaveformAudio {
-  context?: AudioContext;
-  offline?: OfflineAudioContext;
-  analyzer?: AnalyserNode;
-  source?: AudioBufferSourceNode;
-  gain?: GainNode;
-  buffer?: AudioBuffer;
+export class WaveformAudio extends Events<WaveformAudioEvents> {
+  decoder?: AudioDecoder;
+  decoderPromise?: Promise<void>;
+  mediaPromise?: Promise<void>;
+  el?: HTMLAudioElement;
 
+  // private backed by audio element and getters/setters
+  // underscored to keep the public API clean
   private _rate = 1;
   private _volume = 1;
-  private _muted = false;
   private _savedVolume = 1;
-  private _channelCount = 1;
+  private src?: string;
+  private mediaResolve?: () => void;
 
   constructor(options: WaveformAudioOptions) {
-    this.context = this.createAudioContext();
-    this.offline = this.createOfflineAudioContext();
-
-    this._rate = options.rate;
-    this._savedVolume = options.volume;
-    this._volume = options.muted ? 0 : options.volume;
+    super();
+    this._rate = options.rate ?? this._rate;
+    this._savedVolume = options.volume ?? this._volume;
+    this._volume = options.muted ? 0 : this._savedVolume;
+    this.src = options.src;
+    this.createAudioDecoder();
+    this.createMediaElement();
   }
 
   get channelCount() {
-    return this._channelCount;
+    return this.decoder?.channelCount || 1;
+  }
+
+  get duration() {
+    return this.el?.duration ?? 0;
   }
 
   get sampleRate() {
-    return this.context?.sampleRate ?? 44100;
+    return this.decoder?.sampleRate || 44100;
```

Comment:
this is minor but can we create a constant that defines what the 44100 is?

Reply:
yep, good call